Firestore/Source/API/FIRQuery.m

```diff
@@ -256,6 +256,82 @@ - (FIRQuery *)queryWhereFieldPath:(FIRFieldPath *)path isGreaterThanOrEqualTo:(i
+                              "aggregate modifier.");
+    }
+    NSString *path;
+    id value = nil;
+    if (comparison.leftExpression.keyPath == nil &&
+        comparison.rightExpression.keyPath == nil) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate comparison must have "
+                              "a key path on either side of its expression.");
+    } else if (comparison.leftExpression.keyPath == nil) {
+      path = comparison.rightExpression.keyPath;
+      value = comparison.leftExpression.constantValue;
+    } else {
+      path = comparison.leftExpression.keyPath;
+      value = comparison.rightExpression.constantValue;
+    }
+    if (value == nil) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate comparison must have "
+                              "a constant on its expression.");
+    }
+    switch (comparison.predicateOperatorType) {
+      case NSEqualToPredicateOperatorType:
+        return [self queryWhereField:path isEqualTo:value];
+        break;
+      case NSLessThanPredicateOperatorType:
+        return [self queryWhereField:path isLessThan:value];
+        break;
+      case NSLessThanOrEqualToPredicateOperatorType:
+        return [self queryWhereField:path isLessThanOrEqualTo:value];
+        break;
+      case NSGreaterThanPredicateOperatorType:
+        return [self queryWhereField:path isGreaterThan:value];
+        break;
+      case NSGreaterThanOrEqualToPredicateOperatorType:
+        return [self queryWhereField:path isGreaterThanOrEqualTo:value];
+        break;
+      case NSCustomSelectorPredicateOperatorType:
+        FSTThrowInvalidArgument(@"Invalid query. Custom predicate filters are "
+                                "not supported.");
+        break;
+      default:
+        FSTThrowInvalidArgument(@"Invalid query. Predicate does not support "
+                                "%lu operator type.",
+                                (unsigned long)comparison.predicateOperatorType);
+    }
+  } else if ([predicate isKindOfClass:[NSCompoundPredicate class]]) {
+    // Compound predicate
+    NSCompoundPredicate *compound = (NSCompoundPredicate *)predicate;
+    if (compound.compoundPredicateType != NSAndPredicateType ||
+        compound.subpredicates.count == 0) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate only supports "
+                              "AND-compound.");
+    }
+    FIRQuery *query = self;
+    for (NSPredicate *pred in compound.subpredicates) {
+      query = [query queryWherePredicate:pred];
+    }
+    return query;
+  } else if ([predicate isKindOfClass:
+              [[NSPredicate predicateWithBlock:
+                ^BOOL(id obj, NSDictionary *bindings) { return true; }] class]]) {
+    FSTThrowInvalidArgument(@"Invalid query. Block-based predicates are not "
+                            "supported. Please use predicateWithFormat to "
+                            "construct predicate instead.");
```

Comment:
"Construct" is a C++ and Java-ism that's not native to Objective-C (where objects are created or initialized). Also "predicate" should be pluralized.

```
Please use predicateWithFormat to create predicates instead.
```

Here and below.